File: OsmAnd-java/src/main/java/net/osmand/binary/StringBundle.java
package net.osmand.binary;

import net.osmand.util.Algorithms;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import gnu.trove.iterator.TIntObjectIterator;
import gnu.trove.map.hash.TIntObjectHashMap;

public class StringBundle {

	private static final DecimalFormat TWO_DIGITS_FORMATTER = new DecimalFormat("#.##", DecimalFormatSymbols.getInstance(Locale.US));
	private static final DecimalFormat THREE_DIGITS_FORMATTER = new DecimalFormat("#.###", DecimalFormatSymbols.getInstance(Locale.US));
	private static final DecimalFormat FOUR_DIGITS_FORMATTER = new DecimalFormat("#.####", DecimalFormatSymbols.getInstance(Locale.US));
	private static final DecimalFormat FIVE_DIGITS_FORMATTER = new DecimalFormat("#.#####", DecimalFormatSymbols.getInstance(Locale.US));
	private static final DecimalFormat SIX_DIGITS_FORMATTER = new DecimalFormat("#.######", DecimalFormatSymbols.getInstance(Locale.US));

	private Map<String, Item<?>> map = new LinkedHashMap<>();

	public enum ItemType {
		STRING,
		LIST,
		MAP
	}

	public StringBundle() {
	}

	protected StringBundle(Map<String, Item<?>> map) {
		this.map = map;
	}

	public StringBundle newInstance() {
		return new StringBundle();
	}

	public static abstract class Item<T> {

		private final String name;
		private final ItemType type;
		private final T value;

		private Item(String name, ItemType type, T value) {
			this.name = name;
			this.type = type;
			this.value = value;
		}

		public String getName() {
			return name;
		}

		public ItemType getType() {
			return type;
		}

		public T getValue() {
			return value;
		}
	}

	public static class StringItem extends Item<String> {

		private StringItem(String name, String value) {
			super(name, ItemType.STRING, value);
		}

		private StringItem(String name, int value) {
			super(name, ItemType.STRING, String.valueOf(value));
		}

		private StringItem(String name, long value) {
			super(name, ItemType.STRING, String.valueOf(value));
		}

		private StringItem(String name, float value) {
			super(name, ItemType.STRING, String.valueOf(value));
		}

		private StringItem(String name, float value, int maxDigits) {
			super(name, ItemType.STRING, getFormattedValue(value, maxDigits));
		}

		private StringItem(String name, boolean value) {
			super(name, ItemType.STRING, String.valueOf(value));
		}

		private static String getFormattedValue(float value, int maxDigits) {
			DecimalFormat formatter = null;
			if (maxDigits == 2) {
				formatter = TWO_DIGITS_FORMATTER;
			} else if (maxDigits == 3) {
				formatter = THREE_DIGITS_FORMATTER;
			} else if (maxDigits == 4) {
				formatter = FOUR_DIGITS_FORMATTER;
			} else if (maxDigits == 5) {
				formatter = FIVE_DIGITS_FORMATTER;
			} else if (maxDigits == 6) {
				formatter = SIX_DIGITS_FORMATTER;
			}
			return formatter != null ? formatter.format(value) : String.valueOf(value);
		}

		private int asInt(int defaultValue) {
			try {
				return Integer.parseInt(getValue());
			} catch (NumberFormatException e) {
				return defaultValue;
			}
		}

		private long asLong(long defaultValue) {
			try {
				return Long.parseLong(getValue());
			} catch (NumberFormatException e) {
				return defaultValue;
			}
		}

		private float asFloat(float defaultValue) {
			try {
				return Float.parseFloat(getValue());
			} catch (NumberFormatException e) {
				return defaultValue;
			}
		}

		private boolean asBoolean(boolean defaultValue) {
			try {
				return Boolean.parseBoolean(getValue());
			} catch (NumberFormatException e) {
				return defaultValue;
			}
		}

		private int[] asIntArray(int[] defaultValue) {
			try {
				return stringToIntArray(getValue());
			} catch (NumberFormatException e) {
				return defaultValue;
			}
		}

		private int[][] asIntIntArray(int[][] defaultValue) {
			try {
				return stringToIntIntArray(getValue());
			} catch (NumberFormatException e) {
				return defaultValue;
			}
		}
	}

	public static class StringListItem extends Item<List<Item<?>>> {

		private StringListItem(String name, List<Item<?>> list) {
			super(name, ItemType.LIST, list);
		}
	}

	public static class StringMapItem extends Item<Map<String, Item<?>>> {

		private StringMapItem(String name, Map<String, Item<?>> map) {
			super(name, ItemType.MAP, map);
		}
	}

	public static class StringBundleItem extends StringMapItem {

		private StringBundleItem(String name, StringBundle bundle) {
			super(name, bundle.map);
		}
	}

	public Map<String, Item<?>> getMap() {
		return Collections.unmodifiableMap(map);
	}

	public boolean isEmpty() {
		return map.isEmpty();
	}

	public Item<?> getItem(String key) {
		return map.get(key);
	}

	public void putInt(String key, int value) {
		map.put(key, new StringItem(key, value));
	}

	public int getInt(String key, int defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).asInt(defaultValue) : defaultValue;
	}

	public void putLong(String key, long value) {
		map.put(key, new StringItem(key, value));
	}

	public long getLong(String key, long defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).asLong(defaultValue) : defaultValue;
	}

	public void putFloat(String key, float value) {
		map.put(key, new StringItem(key, value));
	}

	public void putFloat(String key, float value, int maxDigits) {
		map.put(key, new StringItem(key, value, maxDigits));
	}

	public float getFloat(String key, float defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).asFloat(defaultValue) : defaultValue;
	}

	public void putBoolean(String key, boolean value) {
		map.put(key, new StringItem(key, value));
	}

	public boolean getBoolean(String key, boolean defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).asBoolean(defaultValue) : defaultValue;
	}

	public void putString(String key, String value) {
		if (value != null) {
			map.put(key, new StringItem(key, value));
		}
	}

	public String getString(String key, String defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).getValue() : defaultValue;
	}

	public void putBundleList(String key, String itemName, List<StringBundle> list) {
		if (list != null) {
			List<Item<?>> itemList = new ArrayList<>();
			for (StringBundle bundle : list) {
				itemList.add(new StringBundleItem(itemName, bundle));
			}
			map.put(key, new StringListItem(key, itemList));
		}
	}

	public void putBundle(String key, StringBundle bundle) {
		map.put(key, new StringBundleItem(key, bundle));
	}

	public void putArray(String key, int[] array) {
		if (array != null) {
			map.put(key, new StringItem(key, intArrayToString(array)));
		}
	}

	public int[] getIntArray(String key, int[] defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).asIntArray(defaultValue) : defaultValue;
	}

	public void putArray(String key, int[][] array) {
		if (array != null) {
			map.put(key, new StringItem(key, intIntArrayToString(array)));
		}
	}

	public int[][] getIntIntArray(String key, int[][] defaultValue) {
		Item<?> item = map.get(key);
		return item instanceof StringItem ? ((StringItem) item).asIntIntArray(defaultValue) : defaultValue;
	}

	public void putArray(String key, long[] array) {
		if (array != null) {
			map.put(key, new StringItem(key, longArrayToString(array)));
		}
	}

	public void putArray(String key, float[] array) {
		if (array != null) {
			map.put(key, new StringItem(key, floatArrayToString(array)));
		}
	}

	public <T> void putMap(String key, TIntObjectHashMap<T> map) {
		if (map != null) {
			StringBundle bundle = newInstance();
			TIntObjectIterator<T> it = map.iterator();
			while (it.hasNext()) {
				it.advance();
				int k = it.key();
				T v = it.value();
				bundle.putString(String.valueOf(k), String.valueOf(v));
			}
			this.map.put(key, new StringBundleItem(key, bundle));
		}
	}

	public <K, V> void putMap(String key, Map<K, V> map) {
		if (map != null) {
			StringBundle bundle = newInstance();
			for (Entry<K, V> entry : map.entrySet()) {
				bundle.putString(String.valueOf(entry.getKey()), String.valueOf(entry.getValue()));
			}
			this.map.put(key, new StringBundleItem(key, bundle));
		}
	}

	private static String intArrayToString(int[] a) {
		if (a == null) {
			return null;
		}
		StringBuilder b = new StringBuilder();
		for (int value : a) {
			if (b.length() > 0) {
				b.append(",");
			}
			b.append(value);
		}
		return b.toString();
	}

	private static int[] stringToIntArray(String a) throws NumberFormatException {
		if (a == null) {
			return null;
		}
		String[] items = a.split(",");
		int[] res = new int[items.length];
		for (int i = 0; i < items.length; i++) {
			res[i] = Integer.parseInt(items[i]);
		}
		return res;
	}

	private static String longArrayToString(long[] a) {
		if (a == null) {
			return null;
		}
		StringBuilder b = new StringBuilder();
		for (long value : a) {
			if (b.length() > 0) {
				b.append(",");
			}
			b.append(value);
		}
		return b.toString();
	}

	private static long[] stringToLongArray(String a) throws NumberFormatException {
		if (a == null) {
			return null;
		}
		String[] items = a.split(",");
		long[] res = new long[items.length];
		for (int i = 0; i < items.length; i++) {
			res[i] = Integer.parseInt(items[i]);
		}
		return res;
	}

	private static String floatArrayToString(float[] a) {
		if (a == null) {
			return null;
		}
		StringBuilder b = new StringBuilder();
		for (float value : a) {
			if (b.length() > 0) {
				b.append(",");
			}
			b.append(value);
		}
		return b.toString();
	}

	private static float[] stringToFloatArray(String a) throws NumberFormatException {
		if (a == null) {
			return null;
		}
		String[] items = a.split(",");
		float[] res = new float[items.length];
		for (int i = 0; i < items.length; i++) {
			res[i] = Float.parseFloat(items[i]);
		}
		return res;
	}

	private static String intIntArrayToString(int[][] a) {
		if (a == null) {
			return null;
		}
		StringBuilder b = new StringBuilder();
		for (int i = 0; i < a.length; i++) {
			if (i > 0) {
				b.append(";");
			}
			int[] arr = a[i];
			if (arr != null && arr.length > 0) {
				b.append(intArrayToString(arr));
			}
		}
		return b.toString();
	}

	private static int[][] stringToIntIntArray(String a) throws NumberFormatException {
		if (a == null) {
			return null;
		}
		String[] items = a.split(";");
		int[][] res = new int[items.length][];
		for (int i = 0; i < items.length; i++) {
			String item = items[i];
			if (Algorithms.isEmpty(item)) {
				res[i] = null;
			} else {
				String[] subItems = item.split(",");
				res[i] = new int[subItems.length];
				for (int k = 0; k < subItems.length; k++) {
					res[i][k] = Integer.parseInt(subItems[k]);
				}
			}
		}
		return res;
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/util/StringBundle.kt
package net.osmand.shared.util

import net.osmand.shared.util.collections.KTIntObjectMap
import kotlin.math.abs
import kotlin.math.floor

open class StringBundle {

	private val map: MutableMap<String, Item<*>>

	constructor() {
		map = LinkedHashMap()
	}

	/** Shares the contents of [bundle], so a subclass can be built around an existing bundle. */
	protected constructor(bundle: StringBundle) {
		this.map = bundle.map
	}

	/** An empty bundle of the same kind, so a reader can build nested bundles of a subclass. */
	open fun newInstance(): StringBundle = StringBundle()

	enum class ItemType {
		STRING,
		LIST,
		MAP
	}

	open class Item<T>(val name: String, val type: ItemType, val value: T?)

	class StringItem(name: String, value: String?) : Item<String>(name, ItemType.STRING, value) {

		constructor(name: String, value: Int) : this(name, value.toString())
		constructor(name: String, value: Long) : this(name, value.toString())
		constructor(name: String, value: Float) : this(name, value.toString())
		constructor(name: String, value: Float, maxDigits: Int) : this(name, formatValue(value, maxDigits))
		constructor(name: String, value: Boolean) : this(name, value.toString())

		fun asInt(defaultValue: Int?): Int? {
			return value?.toIntOrNull() ?: defaultValue
		}

		fun asLong(defaultValue: Long?): Long? {
			return value?.toLongOrNull() ?: defaultValue
		}

		fun asFloat(defaultValue: Float?): Float? {
			return value?.toFloatOrNull() ?: defaultValue
		}

		fun asBoolean(defaultValue: Boolean?): Boolean? {
			// java read this with Boolean.parseBoolean, which answers false for anything but "true"
			return if (value == null) defaultValue else value.equals("true", ignoreCase = true)
		}

		fun asIntArray(defaultValue: IntArray?): IntArray? {
			return try {
				val text = value ?: return defaultValue
				splitDroppingTrailingEmpty(text, ",").map { it.toInt() }.toIntArray()
			} catch (e: NumberFormatException) {
				defaultValue
			}
		}

		fun asIntIntArray(defaultValue: Array<IntArray?>?): Array<IntArray?>? {
			return try {
				val text = value ?: return defaultValue
				val items = splitDroppingTrailingEmpty(text, ";")
				Array(items.size) { i ->
					val item = items[i]
					// an empty item stands for a row that had no values, and reads back as null
					if (item.isEmpty()) {
						null
					} else {
						item.split(",").map { num -> num.toInt() }.toIntArray()
					}
				}
			} catch (e: NumberFormatException) {
				defaultValue
			}
		}

		companion object {

			/**
			 * Splits the way `String.split(regex)` does on the jvm, which drops the empty parts at
			 * the end. The text these arrays are read from was written by that, so the two agree.
			 */
			private fun splitDroppingTrailingEmpty(text: String, separator: String): List<String> {
				if (!text.contains(separator)) {
					// java answers with the whole input when the separator does not occur in it
					return listOf(text)
				}
				val parts = text.split(separator)
				var end = parts.size
				while (end > 0 && parts[end - 1].isEmpty()) {
					end--
				}
				return parts.subList(0, end)
			}

			/**
			 * The same text `DecimalFormat("#.##")` and its wider siblings produced on the jvm:
			 * at most [maxDigits] fraction digits, half even rounding of the exact value, trailing
			 * zeros and a bare decimal point dropped.
			 *
			 * Java only had formatters for two to six digits and fell back to the plain text of the
			 * number otherwise; so does this, and so does a value too large to scale into a Long.
			 */
			private fun formatValue(value: Float, maxDigits: Int): String {
				if (maxDigits < 2 || maxDigits > 6) {
					return value.toString()
				}
				if (value.isNaN()) {
					return "NaN"
				}
				if (value.isInfinite()) {
					return if (value > 0) "\u221E" else "-\u221E"
				}
				val d = value.toDouble()
				if (abs(d) >= 1e12) {
					return value.toString()
				}
				var factor = 1L
				repeat(maxDigits) { factor *= 10L }
				val scaled = abs(d) * factor
				var rounded = floor(scaled)
				val rest = scaled - rounded
				if (rest > 0.5 || (rest == 0.5 && rounded.toLong() % 2L != 0L)) {
					rounded += 1.0
				}
				val units = rounded.toLong()
				val builder = StringBuilder()
				// the sign survives a value that rounds to zero, and so does the sign of -0.0
				if (value.toRawBits() < 0) {
					builder.append('-')
				}
				builder.append(units / factor)
				var fraction = units % factor
				if (fraction != 0L) {
					var digits = maxDigits
					while (fraction % 10L == 0L) {
						fraction /= 10L
						digits--
					}
					val text = fraction.toString()
					builder.append('.')
					repeat(digits - text.length) { builder.append('0') }
					builder.append(text)
				}
				return builder.toString()
			}
		}
	}

	class StringListItem(name: String, list: List<Item<*>>) : Item<List<Item<*>>>(name,
		ItemType.LIST, list)

	open class StringMapItem(name: String, map: Map<String, Item<*>>) : Item<Map<String, Item<*>>>(name,
		ItemType.MAP, map)

	class StringBundleItem(name: String, bundle: StringBundle) : StringMapItem(name, bundle.map)

	fun getMap(): Map<String, Item<*>> {
		return map.toMap()
	}

	fun isEmpty(): Boolean {
		return map.isEmpty()
	}

	fun getItem(key: String): Item<*>? {
		return map[key]
	}

	fun putInt(key: String, value: Int) {
		map[key] = StringItem(key, value)
	}

	fun getInt(key: String, defaultValue: Int): Int {
		return (map[key] as? StringItem)?.asInt(defaultValue) ?: defaultValue
	}

	fun putLong(key: String, value: Long) {
		map[key] = StringItem(key, value)
	}

	fun getLong(key: String, defaultValue: Long?): Long? {
		return (map[key] as? StringItem)?.asLong(defaultValue) ?: defaultValue
	}

	fun putFloat(key: String, value: Float) {
		map[key] = StringItem(key, value)
	}

	fun putFloat(key: String, value: Float, maxDigits: Int) {
		map[key] = StringItem(key, value, maxDigits)
	}

	fun getFloat(key: String, defaultValue: Float?): Float? {
		return (map[key] as? StringItem)?.asFloat(defaultValue) ?: defaultValue
	}

	fun putBoolean(key: String, value: Boolean) {
		map[key] = StringItem(key, value)
	}

	fun getBoolean(key: String, defaultValue: Boolean?): Boolean? {
		return (map[key] as? StringItem)?.asBoolean(defaultValue) ?: defaultValue
	}

	fun putString(key: String, value: String?) {
		map[key] = StringItem(key, value)
	}

	fun getString(key: String, defaultValue: String?): String? {
		return (map[key] as? StringItem)?.value ?: defaultValue
	}

	fun putBundleList(key: String, itemName: String, list: List<StringBundle>?) {
		list?.let {
			val itemList = it.map { bundle -> StringBundleItem(itemName, bundle) }
			map[key] = StringListItem(key, itemList)
		}
	}

	fun putBundle(key: String, bundle: StringBundle) {
		map[key] = StringBundleItem(key, bundle)
	}

	fun putArray(key: String, array: IntArray?) {
		array?.let {
			map[key] = StringItem(key, it.joinToString(","))
		}
	}

	fun getIntArray(key: String, defaultValue: IntArray?): IntArray? {
		return (map[key] as? StringItem)?.asIntArray(defaultValue) ?: defaultValue
	}

	/** Rows may be null or empty, and are written as an empty item, the way the java bundle did. */
	fun putArray(key: String, array: Array<IntArray?>?) {
		array?.let { rows ->
			map[key] = StringItem(key, rows.joinToString(";") { row ->
				if (row == null) "" else row.joinToString(",")
			})
		}
	}

	fun getIntIntArray(key: String, defaultValue: Array<IntArray?>?): Array<IntArray?>? {
		return (map[key] as? StringItem)?.asIntIntArray(defaultValue) ?: defaultValue
	}

	fun putArray(key: String, array: LongArray?) {
		array?.let {
			map[key] = StringItem(key, it.joinToString(","))
		}
	}

	fun putArray(key: String, array: FloatArray?) {
		array?.let {
			map[key] = StringItem(key, it.joinToString(","))
		}
	}

	fun <T : Any> putMap(key: String, map: KTIntObjectMap<T>) {
		val bundle = StringBundle()
		map.forEach { k, v -> bundle.putString(k.toString(), v.toString()) }
		this.map[key] = StringBundleItem(key, bundle)
	}

	fun <K, V> putMap(key: String, map: Map<K, V>) {
		val bundle = StringBundle()
		map.forEach { (k, v) -> bundle.putString(k.toString(), v.toString()) }
		this.map[key] = StringBundleItem(key, bundle)
	}
}
